File: nvcs/classifier.py
'''
Created on Mar 22, 2013

@author: knimerfro
'''

class Node:
    def __init__(self, ident, parent, match_function, description):
        self.ident = ident
        self.parent = parent
        self.match_function = match_function
        self.description = description
        self.children = []

    def __repr__(self):
        return "Node(%r, %r, %r, %r)" % (self.ident, self.parent, self.match_function, self.description)

class Solution:
    def __init__(self, obj, path):
        self.obj = obj
        self.path = path
        
    def __repr__(self):
        return "Solution(%r, %r)" % (self.obj, self.path)

class Classifier:  
    def __init__(self, key):
        self.root_node = key.root_node

    def __repr__(self):
        return "Classifier(%r, %r)" % (self.root_node)

    def choose(self, nodes, obj):
        for node in nodes:
            if node.match_function(obj):
                return node
        return None

    def walk(self, root_node, obj):
        node = root_node
        while node is not None and node.children:
            node = self.choose(node.children, obj)
            yield node

    def classify(self, obj):
        path = [node for node in self.walk(self.root_node, obj)]
        return Solution(obj, path)

File: nvcs-dev/nvcs-java/Classifier.java
import java.util.ArrayList;
import java.util.List;

public class Classifier {

    private Object root_node;

    public Classifier(Object key) {
        root_node = key;
    }

    @Override
    public String toString() {
        return String.format("Classifier(%s)", root_node);
    }

    public Object choose(Object[] nodes, Object obj) {
        for (Object node: nodes) {
            if (node.match_function(obj)) {
                return node;
            }
        }
        return null;
    }

    // public Object walk(Object root_node, Object obj) {
    //     Object node = root_node;
    //     while (node != null && node.children != null) {
    //         node = choose(node.children, obj);
    //         return node;
    //     }
    // }

    public Solution classify(Object obj) {
        List<Object> path = new ArrayList<>();
        Object node = root_node;
        while (node != null && node.children != null) {
            node = choose(node.children, obj);
            path.add(node);
        }
        return Solution(obj, path);
    }
}
